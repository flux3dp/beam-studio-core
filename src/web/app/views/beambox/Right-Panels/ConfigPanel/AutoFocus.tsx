import classNames from 'classnames';
import React, { memo, useContext } from 'react';

import UnitInput from 'app/widgets/Unit-Input-v2';
import useI18n from 'helpers/useI18n';
import { CUSTOM_PRESET_CONSTANT, DataType, writeData } from 'helpers/layer/layer-config-helper';

import ConfigPanelContext from './ConfigPanelContext';
import styles from './Block.module.scss';

const AutoFocus = (): JSX.Element => {
  const lang = useI18n();
  const t = lang.beambox.right_panel.laser_panel;
  const { selectedLayers, state, dispatch } = useContext(ConfigPanelContext);
  const { height, repeat, zStep } = state;

  const handleToggle = () => {
    const value = -height.value;
    dispatch({ type: 'change', payload: { height: value } });
    selectedLayers.forEach((layerName) => writeData(layerName, DataType.height, value));
  };

  const handleHeightChange = (value: number) => {
    dispatch({ type: 'change', payload: { height: value } });
    selectedLayers.forEach((layerName) => writeData(layerName, DataType.height, value));
  };

  const handleZStepChange = (value: number) => {
    dispatch({ type: 'change', payload: { zStep: value, configName: CUSTOM_PRESET_CONSTANT } });
    selectedLayers.forEach((layerName) => {
      writeData(layerName, DataType.zstep, value);
      writeData(layerName, DataType.configName, CUSTOM_PRESET_CONSTANT);
    });
  };

  return (
    <>
      <div className={classNames(styles.panel, styles.checkbox)} onClick={handleToggle}>
        <span className={styles.title}>{t.focus_adjustment}</span>
        <input type="checkbox" checked={height.value > 0} readOnly />
      </div>
      {height.value > 0 ? (
        <div className={classNames(styles.panel, styles['without-drag'])}>
          <span className={styles.title}>{t.height}</span>
          <UnitInput
            id="height"
            className={{ [styles.input]: true }}
            min={0.01}
            max={20}
            unit="mm"
            defaultValue={height.value}
            getValue={handleHeightChange}
            displayMultiValue={height.hasMultiValue}
          />
        </div>
      ) : null}
      {repeat.value > 1 && height.value > 0 ? (
        <div className={classNames(styles.panel, styles['without-drag'])}>
          <span className={styles.title}>{t.z_step}</span>
          <UnitInput
            id="z_step"
            className={{ [styles.input]: true }}
            min={0}
            max={20}
            unit="mm"
            defaultValue={zStep.value}
            getValue={handleZStepChange}
            displayMultiValue={zStep.hasMultiValue}
          />
        </div>
      ) : null}
    </>
  );
};

export default memo(AutoFocus);
