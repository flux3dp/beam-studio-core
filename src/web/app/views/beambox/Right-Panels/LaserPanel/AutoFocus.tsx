import classNames from 'classnames';
import React from 'react';

import UnitInput from 'app/widgets/Unit-Input-v2';
import useI18n from 'helpers/useI18n';

import styles from './AutoFocus.module.scss';

interface Props {
  height: number;
  hasMultiHeight: boolean;
  repeat: number;
  zStep: number;
  hasMultiZStep: boolean;
  onToggle: () => void;
  onHeightChange: (val: number) => void;
  onZStepChange: (val: number) => void;
}

const AutoFocus = ({
  height,
  hasMultiHeight,
  repeat,
  zStep,
  hasMultiZStep,
  onToggle,
  onHeightChange,
  onZStepChange,
}: Props): JSX.Element => {
  const lang = useI18n();
  const t = lang.beambox.right_panel.laser_panel;

  return (
    <>
      <div className={classNames(styles.panel, styles.checkbox)} onClick={onToggle}>
        <span className={styles.title}>{t.focus_adjustment}</span>
        <input type="checkbox" checked={height > 0} readOnly />
      </div>
      {height > 0 ? (
        <div className={classNames(styles.panel, styles['without-drag'])}>
          <span className={styles.title}>{t.height}</span>
          <UnitInput
            id="height"
            className={{ [styles.input]: true }}
            min={0.01}
            max={20}
            unit="mm"
            defaultValue={height}
            getValue={onHeightChange}
            displayMultiValue={hasMultiHeight}
          />
        </div>
      ) : null}
      {repeat > 1 && height > 0 ? (
        <div className={classNames(styles.panel, styles['without-drag'])}>
          <span className={styles.title}>{t.z_step}</span>
          <UnitInput
            id="z_step"
            className={{ [styles.input]: true }}
            min={0}
            max={20}
            unit="mm"
            defaultValue={zStep}
            getValue={onZStepChange}
            displayMultiValue={hasMultiZStep}
          />
        </div>
      ) : null}
    </>
  );
};

export default AutoFocus;
