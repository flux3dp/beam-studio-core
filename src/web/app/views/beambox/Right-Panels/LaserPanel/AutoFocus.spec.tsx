import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import AutoFocus from './AutoFocus';

jest.mock('helpers/useI18n', () => () => ({
  beambox: {
    right_panel: {
      laser_panel: {
        focus_adjustment: 'focus_adjustment',
        height: 'height',
        z_step: 'z_step',
      },
    },
  },
}));

jest.mock('app/widgets/Unit-Input-v2', () => (
  { id, min, max, unit, defaultValue, decimal, displayMultiValue, getValue }: any
) => (
  <div>
    MockUnitInput
    <p>id: {id}</p>
    <p>min: {min}</p>
    <p>max: {max}</p>
    <p>unit: {unit}</p>
    <p>defaultValue: {defaultValue}</p>
    <p>decimal: {decimal}</p>
    <p>displayMultiValue: {displayMultiValue}</p>
    <button type="button" onClick={() => getValue(7)}>change-{id}</button>
  </div>
));

const mockOnToggle = jest.fn();
const mockOnHeightChange = jest.fn();
const mockOnZStepChange = jest.fn();

describe('test AutoFocus', () => {
  it('should render correctly when height is less than 0', () => {
    const { container, queryByText } = render(
      <AutoFocus
        height={-3}
        hasMultiHeight={false}
        repeat={1}
        zStep={0}
        hasMultiZStep={false}
        onToggle={mockOnToggle}
        onHeightChange={mockOnHeightChange}
        onZStepChange={mockOnZStepChange}
      />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText('height')).not.toBeInTheDocument();
    expect(queryByText('z_step')).not.toBeInTheDocument();
  });

  it('should render correctly when repeat is less than 1', () => {
    const { container, queryByText } = render(
      <AutoFocus
        height={3}
        hasMultiHeight={false}
        repeat={1}
        zStep={0}
        hasMultiZStep={false}
        onToggle={mockOnToggle}
        onHeightChange={mockOnHeightChange}
        onZStepChange={mockOnZStepChange}
      />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText('height')).toBeInTheDocument();
    expect(queryByText('z_step')).not.toBeInTheDocument();
  });

  it('should render correctly when repeat is larger than 1', () => {
    const { container, queryByText } = render(
      <AutoFocus
        height={3}
        hasMultiHeight={false}
        repeat={2}
        zStep={0}
        hasMultiZStep={false}
        onToggle={mockOnToggle}
        onHeightChange={mockOnHeightChange}
        onZStepChange={mockOnZStepChange}
      />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText('height')).toBeInTheDocument();
    expect(queryByText('z_step')).toBeInTheDocument();
  });

  test('handlers should work', () => {
    const { container, getByText } = render(
      <AutoFocus
        height={3}
        hasMultiHeight={false}
        repeat={2}
        zStep={0}
        hasMultiZStep={false}
        onToggle={mockOnToggle}
        onHeightChange={mockOnHeightChange}
        onZStepChange={mockOnZStepChange}
      />
    );
    expect(mockOnToggle).not.toBeCalled();
    fireEvent.click(container.querySelector('.checkbox'));
    expect(mockOnToggle).toBeCalledTimes(1);

    expect(mockOnHeightChange).not.toBeCalled();
    fireEvent.click(getByText('change-height'));
    expect(mockOnHeightChange).toBeCalledTimes(1);
    expect(mockOnHeightChange).lastCalledWith(7);

    expect(mockOnZStepChange).not.toBeCalled();
    fireEvent.click(getByText('change-z_step'));
    expect(mockOnZStepChange).toBeCalledTimes(1);
    expect(mockOnZStepChange).lastCalledWith(7);
  });
});
