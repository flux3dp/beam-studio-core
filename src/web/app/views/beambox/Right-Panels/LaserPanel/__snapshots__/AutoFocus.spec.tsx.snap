// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AutoFocus should render correctly when height is less than 0 1`] = `
<div>
  <div
    class="panel checkbox"
  >
    <span
      class="title"
    >
      focus_adjustment
    </span>
    <input
      readonly=""
      type="checkbox"
    />
  </div>
</div>
`;

exports[`test AutoFocus should render correctly when repeat is larger than 1 1`] = `
<div>
  <div
    class="panel checkbox"
  >
    <span
      class="title"
    >
      focus_adjustment
    </span>
    <input
      checked=""
      readonly=""
      type="checkbox"
    />
  </div>
  <div
    class="panel without-drag"
  >
    <span
      class="title"
    >
      height
    </span>
    <div>
      MockUnitInput
      <p>
        id: 
        height
      </p>
      <p>
        min: 
        0.01
      </p>
      <p>
        max: 
        20
      </p>
      <p>
        unit: 
        mm
      </p>
      <p>
        defaultValue: 
        3
      </p>
      <p>
        decimal: 
      </p>
      <p>
        displayMultiValue: 
      </p>
      <button
        type="button"
      >
        change-
        height
      </button>
    </div>
  </div>
  <div
    class="panel without-drag"
  >
    <span
      class="title"
    >
      z_step
    </span>
    <div>
      MockUnitInput
      <p>
        id: 
        z_step
      </p>
      <p>
        min: 
        0
      </p>
      <p>
        max: 
        20
      </p>
      <p>
        unit: 
        mm
      </p>
      <p>
        defaultValue: 
        0
      </p>
      <p>
        decimal: 
      </p>
      <p>
        displayMultiValue: 
      </p>
      <button
        type="button"
      >
        change-
        z_step
      </button>
    </div>
  </div>
</div>
`;

exports[`test AutoFocus should render correctly when repeat is less than 1 1`] = `
<div>
  <div
    class="panel checkbox"
  >
    <span
      class="title"
    >
      focus_adjustment
    </span>
    <input
      checked=""
      readonly=""
      type="checkbox"
    />
  </div>
  <div
    class="panel without-drag"
  >
    <span
      class="title"
    >
      height
    </span>
    <div>
      MockUnitInput
      <p>
        id: 
        height
      </p>
      <p>
        min: 
        0.01
      </p>
      <p>
        max: 
        20
      </p>
      <p>
        unit: 
        mm
      </p>
      <p>
        defaultValue: 
        3
      </p>
      <p>
        decimal: 
      </p>
      <p>
        displayMultiValue: 
      </p>
      <button
        type="button"
      >
        change-
        height
      </button>
    </div>
  </div>
</div>
`;
